resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/AdaptiveBulkhead.java

```diff
@@ -0,0 +1,546 @@
+	 */
+	static <T> CheckedConsumer<T> decorateCheckedConsumer(AdaptiveBulkhead bulkhead, CheckedConsumer<T> consumer) {
+		return (t) -> {
+			bulkhead.acquirePermission();
+			Instant start = null;
+			try {
+				start = Instant.now();
+				consumer.accept(t);
+
+			} catch (Exception e) {
+				handleError(bulkhead, start, e);
+				throw e;
+			} finally {
+				if (start != null) {
+					Instant finish = Instant.now();
+					bulkhead.onComplete(Duration.between(start, finish), true);
+				}
+			}
+		};
+	}
+
+	/**
+	 * Returns a runnable which is decorated by a bulkhead.
+	 *
+	 * @param bulkhead the bulkhead
+	 * @param runnable the original runnable
+	 * @return a runnable which is decorated by a bulkhead.
+	 */
+	static Runnable decorateRunnable(AdaptiveBulkhead bulkhead, Runnable runnable) {
+		return () -> {
+			bulkhead.acquirePermission();
+			Instant start = null;
+			try {
+				start = Instant.now();
+				runnable.run();
+			} catch (Exception e) {
+				handleError(bulkhead, start, e);
+				throw e;
+			} finally {
+				if (start != null) {
+					Instant finish = Instant.now();
+					bulkhead.onComplete(Duration.between(start, finish), true);
+				}
+			}
+		};
+	}
+
+	/**
+	 * Returns a function which is decorated by a bulkhead.
+	 *
+	 * @param bulkhead the bulkhead
+	 * @param function the original function
+	 * @param <T>      the type of the input to the function
+	 * @param <R>      the type of the result of the function
+	 * @return a function which is decorated by a bulkhead.
+	 */
+	static <T, R> Function<T, R> decorateFunction(AdaptiveBulkhead bulkhead, Function<T, R> function) {
+		return (T t) -> {
+			bulkhead.acquirePermission();
+			Instant start = null;
+			try {
+				start = Instant.now();
+				return function.apply(t);
+			} catch (Exception e) {
+				handleError(bulkhead, start, e);
+				throw e;
+			} finally {
+				if (start != null) {
+					Instant finish = Instant.now();
+					bulkhead.onComplete(Duration.between(start, finish), true);
+				}
+			}
+		};
+	}
+
+	/**
+	 * Returns a function which is decorated by a bulkhead.
+	 *
+	 * @param bulkhead the bulkhead
+	 * @param function the original function
+	 * @param <T>      the type of the input to the function
+	 * @param <R>      the type of the result of the function
+	 * @return a function which is decorated by a bulkhead.
+	 */
+	static <T, R> CheckedFunction1<T, R> decorateCheckedFunction(AdaptiveBulkhead bulkhead, CheckedFunction1<T, R> function) {
+		return (T t) -> {
+			bulkhead.acquirePermission();
+			Instant start = null;
+			try {
+				start = Instant.now();
+				return function.apply(t);
+			} catch (Exception e) {
+				handleError(bulkhead, start, e);
+				throw e;
+			} finally {
+				if (start != null) {
+					Instant finish = Instant.now();
+					bulkhead.onComplete(Duration.between(start, finish), true);
+				}
+			}
+		};
+	}
+
+	static void handleError(AdaptiveBulkhead bulkhead, @Nullable Instant start, Exception e) {
```

Comment:
Can we move this into an `onError` method` of the AdaptiveBulkhead implementation.

Reply:
why ? handleError is needed into the default methods of Adapive bulkhead interface